nixos/modules/programs/nix-required-mounts.nix

```diff
@@ -0,0 +1,85 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.programs.nix-required-mounts;
+  hook =
+    pkgs.nix-required-mounts.override { inherit (cfg) allowedPatterns; };
+
+  patternType = with lib.types; submodule ({ config, name, ... }: {
+    options.onFeatures = lib.mkOption {
+      type = listOf str;
+      description = "Which requiredSystemFeatures should trigger relaxation of the sandbox";
+      default = [ name ];
+    };
+    options.paths = lib.mkOption {
+      type = listOf path;
+      description = "A list of glob patterns, indicating which paths to expose to the sandbox";
+    };
+  });
+
+  defaults = {
+    opengl.onFeatures = [ "opengl" ];
+    opengl.paths = [
+      "/dev/video*"
+      "/dev/dri"
+
+      pkgs.addOpenGLRunpath.driverLink
+      # /run/opengl-driver/lib only contains symlinks
+      config.hardware.opengl.package
+    ] ++ config.hardware.opengl.extraPackages;
+    cuda.onFeatures = [ "cuda" ];
+    cuda.paths = defaults.opengl.paths ++ [ "/dev/nvidia*" ];
+  };
+in
+{
+  meta.maintainers = with lib.maintainers; [ SomeoneSerge ];
+  options.programs.nix-required-mounts = {
+    enable = lib.mkEnableOption
+      "Expose extra paths to the sandbox depending on derivations' requiredSystemFeatures";
+    presets.opengl.enable = lib.mkOption {
+      type = lib.types.bool;
+      default = config.hardware.opengl.enable;
+      defaultText = lib.literalExpression "hardware.opengl.enable";
+      description = ''
+        Expose OpenGL drivers to derivations marked with requiredSystemFeatures = [ "opengl" ]
+      '';
+    };
+    presets.cuda.enable = lib.mkEnableOption ''
+      Expose CUDA drivers and GPUs to derivations marked with requiredSystemFeatures = [ "cuda" ]
+    '';
+    allowedPatterns = with lib.types;
+      lib.mkOption rec {
+        type = attrsOf patternType;
```

Comment:
@SomeoneSerge in my experience, `attrsOf` and `listOf` and friends expect their argument to be a type -- I expected to see `patternType.type` here instead.

As an example:

https://github.com/NixOS/nixpkgs/blob/157b25b88712903f31d159db55f56d91331f141f/pkgs/development/cuda-modules/modules/generic/manifests/feature/release.nix#L65

Reply:
`patternType` is a type, it's an instance of `submodule`

Reply:
Oh! That makes more sense -- thanks!